### DISCUSSION TITLE
Unable to connect to Ollama

### DISCUSSION THREAD

**Discussion by galsi** [open-webui/open-webui#2849] [General]
# Bug Report

## Description
Cannot select model in main drop-down or in admin panel .

**Bug Summary:**
Cannot select model on any of the select options

**Steps to Reproduce:**
installed on clean machine 
Ubuntu 24.04
open-webui
ollama with GPU support
Tested ollama without webui - working.
Tested Openwebui on firefox and Chrome both cannot select model

**Expected Behavior:**
select a model 

**Actual Behavior:**
model is not selected or displayed
model is listed in the models list and connection is varified

## Environment

- **Open WebUI Version:** [v0.2.4]
- **Ollama (if applicable):** [ollama version is 0.1.41]

- **Operating System:** [Ubuntu 24.04]
- **Browser (if applicable):** [Chrome 125.0.6422.141, Firefox 126.01]

## Reproduction Details

**Confirmation:**

- [X ] I have read and followed all the instructions provided in the README.md.
- [ X] I am on the latest version of both Open WebUI and Ollama.
- [ ] I have included the browser console logs.
- [ X] I have included the Docker container logs.

## Logs and Screenshots

**Browser Console Logs:**
[Include relevant browser console logs, if applicable]
![Screenshot from 2024-06-05 17-07-38](https://github.com/open-webui/open-webui/assets/16662563/cef87942-5917-4b55-b41d-6e0b222a2fa5)

**Docker Container Logs:**
```
Loading WEBUI_SECRET_KEY from file, not provided as an environment variable.
Generating WEBUI_SECRET_KEY
Loading WEBUI_SECRET_KEY from .webui_secret_key
INFO:     Started server process [1]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
/app

  ___                    __        __   _     _   _ ___ 
 / _ \ _ __   ___ _ __   \ \      / /__| |__ | | | |_ _|
| | | | '_ \ / _ \ '_ \   \ \ /\ / / _ \ '_ \| | | || | 
| |_| | |_) |  __/ | | |   \ V  V /  __/ |_) | |_| || | 
 \___/| .__/ \___|_| |_|    \_/\_/ \___|_.__/ \___/|___|
      |_|                                               

      
v0.2.4 - building the best open-source AI user interface.

https://github.com/open-webui/open-webui

INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:     172.17.0.1:51944 - "POST /ollama/urls/update HTTP/1.1" 401 Unauthorized
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:     172.17.0.1:51944 - "POST /ollama/urls/update HTTP/1.1" 401 Unauthorized
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:     172.17.0.1:51944 - "POST /ollama/urls/update HTTP/1.1" 401 Unauthorized
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:     172.17.0.1:51944 - "GET / HTTP/1.1" 304 Not Modified
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:     172.17.0.1:51944 - "GET /api/config HTTP/1.1" 200 OK
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:     172.17.0.1:51944 - "GET /api/v1/auths/ HTTP/1.1" 401 Unauthorized
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:     172.17.0.1:51944 - "GET /static/favicon.png HTTP/1.1" 200 OK
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:apps.webui.models.auths:authenticate_user: gals@camtek.com
INFO:     172.17.0.1:51944 - "POST /api/v1/auths/signin HTTP/1.1" 200 OK
INFO:apps.ollama.main:get_all_models()
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:     172.17.0.1:51944 - "GET /api/changelog HTTP/1.1" 200 OK
INFO:     172.17.0.1:57412 - "GET /api/v1/users/user/settings HTTP/1.1" 200 OK
INFO:apps.ollama.main:get_all_models()
INFO:apps.ollama.main:get_all_models()
INFO:apps.ollama.main:get_all_models()
INFO:apps.ollama.main:get_all_models()
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:     172.17.0.1:57420 - "GET /api/v1/configs/banners HTTP/1.1" 200 OK
INFO:     172.17.0.1:51944 - "GET /api/models HTTP/1.1" 200 OK
INFO:     172.17.0.1:57412 - "GET /api/v1/prompts/ HTTP/1.1" 200 OK
INFO:     172.17.0.1:57418 - "GET /api/v1/documents/ HTTP/1.1" 200 OK
INFO:     172.17.0.1:57426 - "GET /api/v1/chats/tags/all HTTP/1.1" 200 OK
INFO:apps.ollama.main:get_all_models()
INFO:apps.ollama.main:get_all_models()
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
ERROR:apps.ollama.main:Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Connection refused]
INFO:apps.ollama.main:get_all_models()

```

**Screenshots (if applicable):**
[Attach any relevant screenshots to help illustrate the issue]

## Installation Method

[ Docker]

## Additional Information

Firewall Status : Inactive

## Note

If the bug report is incomplete or does not follow the provided instructions, it may not be addressed. Please ensure that you have followed the steps outlined in the README.md and troubleshooting.md documents, and provide all necessary information for us to reproduce and address the issue. Thank you!

**Comment by justinh-rahb** [comment]
Set `OLLAMA_HOST=0.0.0.0` and `OLLAMA_ORIGINS=*`

https://github.com/ollama/ollama/blob/main/docs/faq.md#setting-environment-variables-on-linux

**Reply by regiellis** [reply]
Is this not a huge security risk? I know the team is recommending this because I believe the Ollama team is...but would it not be better to either reverse proxy the local address via something like Caddy or in my case use docker's bridge IP in place of 0.0.0.0: (http://172.17.0.1:11434)

**Reply by justinh-rahb** [reply]
Indeed it would be better to be more specific in the interface you bind to than `0.0.0.0`, and the fact that you know enough to ask... I'd highly encourage _you_ to do so. However it's been a struggle for us to help the users that don't understand why this is even necessary, and we've preferred to take the path of least resistance.

**Reply by Epannec** [reply]
Is it possible to elaborate a bit on this fix? I don't quite understand how a setting in the ollama container would allow webui (which runs as a seperate podman container on my system) to change url for the ollama API? This used to work perfectly with ollama-webui. 

I would like to point towards the Ollama instance as if it was running on a different system. This used to be possible using OLLAMA_BASE_URL which is still being referred to in the README), but now seems to be ignored?

Thanks for helping me understand,
Erwin

**Reply by justinh-rahb** [reply]
That is already the default setting for Ollama's Docker image. It would not present a security risk if you do not publish the port to all interfaces, and only allowed it to be accessed from within the stack network. If you require any further explanation of Docker networking, I suggest reading their documentation. If you need more information on networking in general, I suggest maybe picking up a CCNA study guide.

**Reply by Epannec** [reply]
Oke, I understand the access policy for the ollama API, and I'm absolutely fine by it.

What I do not understand, is how the access policy of ollama should enable/disable setting the Ollama API in the open-webui. I previously was able to set this using the OLLAMA_BASE_URL environment variable, but this is not being used anymore and I cannot change the setting in the webinterface, as shown above.

In other words: how can I change the url for ollama API from "http://host.docker.internal" to another url?

Thanks,
Erwin

**Reply by iomatix** [reply]
Not sure if it will resolve the problem but for Podman, remove the container and setup a new container with this.
`podman run -d -p 3000:8080 --add-host=host.containers.internal:host-gateway -v open-webui:/app/backend/data --name open-webui --restart always ghcr.io/open-webui/open-webui:main `
`host.docker.internal` is changed to `host.containers.internal`

**Comment by rutvik2611** [comment]
I tried setting the environment variables as follows:

- `OLLAMA_HOST=0.0.0.0`
- `OLLAMA_ORIGINS=*`

However, I'm still encountering a connection error. Below is the `docker-compose.yml` code I am using:

```yaml
version: "3.3"

services:
  open-webui:
    image: ghcr.io/open-webui/open-webui:main
    container_name: open-webui
    restart: always
    # network_mode: host  # Uncomment this line to use host network if needed

    ports:
      - "11434:11434"   # Forward host port 11434 to container port 11434
      - "8088:8080"     # Forward host port 8088 to container port 8080
    volumes:
      - ./persist/open-webui:/app/backend/data
      - /var/run/docker.sock:/var/run/docker.sock # Pass Docker socket into the container
    environment:
      - OLLAMA_BASE_URL=http://127.0.0.1:11434 # Use service name for internal communication
      - OLLAMA_HOST=0.0.0.0
      - OLLAMA_ORIGINS=*

    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.open-webui-http.rule=Host(`open-webui.${MAIN_DOMAIN}`)"
      - "traefik.http.routers.open-webui-http.entrypoints=web"
      - "traefik.http.routers.open-webui-secure.rule=Host(`open-webui.${MAIN_DOMAIN}`)"
      - "traefik.http.routers.open-webui-secure.entrypoints=websecure"
      - "traefik.http.routers.open-webui-secure.tls.certresolver=myresolver"
      - "traefik.http.services.open-webui.loadbalancer.server.port=8080"
    networks:
      backboneNetwork:
        ipv4_address: 172.20.0.11

networks:
  backboneNetwork:
    external: true
    ipam:
      config:
        - subnet: "172.20.0.0/16"
'''
Error
'''
INFO:     174.49.73.31:57459 - "GET /api/models HTTP/1.1" 200 OK
ERROR [apps.ollama.main] Connection error: Cannot connect to host host.docker.internal:11434 ssl:default [Name or service not known]
INFO:     174.49.73.31:57462 - "GET /api/v1/documents/ HTTP/1.1" 200 OK
'''
